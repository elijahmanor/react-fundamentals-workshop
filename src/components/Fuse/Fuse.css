.Fuse {
  background-color: var(--color-section-bg);
  border-bottom: 1px solid var(--color-divider);
  display: flex;
}

.Fuse-avatar {
  border-radius: 50%;
  height: 4em;
  width: 4em;
  margin: 1rem;
  max-width: 100%;
}

.Fuse-body {
  width: 100%;
  margin: 1rem 1rem 1rem 0;
}

.Fuse-header {
  display: flex;
  color: var(--color-text-secondary);
}

.Fuse-fullName {
  color: var(--color-text-primary);
  font-weight: bold;
}

.Fuse-message {
  margin-bottom: 0.5rem;
}

.Icon {
  background: none;
  color: inherit;
  border: none;
  padding: 0;
  font: inherit;
  cursor: pointer;
  outline: inherit;
  margin-right: 1rem;
}

.Icon .fa.is-active {
  color: #ed1b2e;
}
