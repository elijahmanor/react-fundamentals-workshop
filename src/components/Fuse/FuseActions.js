import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

import "./Fuse.css";

const FuseActions = ({ id, userName, bomb, onReply, onBomb }) => {
  const classes = classNames("fa fa-bomb", {
    "is-active": bomb
  });
  onBomb = onBomb.bind(null, { id, bomb: !bomb });
  return (
    <div className="FuseActions">
      <button className="FuseActions-action" onClick={onReply}>
        <i className="fa fa-comment-o" aria-hidden="true" />
      </button>
      <button className="FuseActions-action is-active" onClick={onBomb}>
        <i className={classes} aria-hidden="true" />
      </button>
    </div>
  );
};

FuseActions.propTypes = {
  id: PropTypes.number.isRequired,
  userName: PropTypes.string.isRequired,
  bomb: PropTypes.bool.isRequired,
  onReply: PropTypes.func,
  onBomb: PropTypes.func
};

FuseActions.defaultProps = {
  onReply: () => {},
  onBomb: () => {}
};

export default FuseActions;
