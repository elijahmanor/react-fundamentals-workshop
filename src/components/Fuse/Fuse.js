import React from "react";
import md5 from "md5";
import { format, formatDistanceStrict } from "date-fns";
import PropTypes from "prop-types";
import FuseActions from "./FuseActions";

import "./Fuse.css";

const Fuse = ({ id, date, userName, fullName, avatar, email, message, bomb, onBomb, onReply }) => {
  const url = avatar || `https://www.gravatar.com/avatar/${md5(email)}`;
  const user = { userName, fullName };

  return (
    <div className="Fuse">
      <img className="Fuse-avatar" alt={`${fullName}'s Avatar`} src={url} />
      <article className="Fuse-body">
        <header className="Fuse-header">
          <div className="Fuse-fullName">{fullName}</div>
          <span>&nbsp;</span>
          <div className="Fuse-userName">@{userName}</div>
          <span>&nbsp;â€”&nbsp;</span>
          <div className="Fuse-date" title={format(date, "LLL")}>
            {formatDistanceStrict(date, new Date())}
          </div>
        </header>
        <div className="Fuse-message">{message}</div>
        <FuseActions id={id} user={user} bomb={bomb} onBomb={onBomb} onReply={onReply} />
      </article>
    </div>
  );
};

Fuse.propTypes = {
  id: PropTypes.string.isRequired,
  date: PropTypes.instanceOf(Date).isRequired,
  userName: PropTypes.string.isRequired,
  fullName: PropTypes.string.isRequired,
  avatar: PropTypes.string,
  email: PropTypes.string,
  message: PropTypes.string.isRequired,
  bomb: PropTypes.bool.isRequired,
  onBomb: PropTypes.func.isRequired,
  onReply: PropTypes.func.isRequired
};

React.defaultProps = {
  onBomb: () => {}
};

export default Fuse;
