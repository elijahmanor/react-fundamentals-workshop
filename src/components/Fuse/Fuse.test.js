import React from "react";
import { shallow, mount } from "enzyme";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import Fuse from "./Fuse";

describe("Fuse", () => {
  beforeEach(() => {
    const NOW = new Date("2017-10-10T00:00:00");
    global.Date = jest.fn(() => NOW);
  });

  it("renders with an avatar", () => {
    const tree = renderer.create(
      <Fuse
        id={1}
        date={new Date("2017-10-10T10:10:10")}
        userName="johnsmith"
        fullName="John Smith"
        avatar="http://avatar.com/me.png"
        message="Hello World!"
        favorite={false}
      />
    );
    expect(tree).toMatchSnapshot();
  });

  it("renders with an email", () => {
    const tree = renderer.create(
      <Fuse
        id={1}
        date={new Date("2017-10-10T10:10:10")}
        userName="johnsmith"
        fullName="John Smith"
        avatar="http://avatar.com/me.png"
        message="Hello World!"
        favorite={false}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
