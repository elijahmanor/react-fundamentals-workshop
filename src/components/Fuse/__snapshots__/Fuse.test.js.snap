// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fuse renders with an avatar 1`] = `
<div
  className="Fuse"
>
  <img
    alt="John Smith's Avatar"
    className="Fuse-avatar"
    src="http://avatar.com/me.png"
  />
  <article
    className="Fuse-body"
  >
    <header
      className="Fuse-header"
    >
      <div
        className="Fuse-fullName"
      >
        John Smith
      </div>
      <span>
         
      </span>
      <div
        className="Fuse-userName"
      >
        @
        johnsmith
      </div>
      <span>
         — 
      </span>
      <div
        className="Fuse-date"
        title="October 9 2017 7:00 p.m."
      >
        0 seconds
      </div>
    </header>
    <div
      className="Fuse-message"
    >
      Hello World!
    </div>
    <div
      className="FuseActions"
    >
      <button
        className="FuseActions-action"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="fa fa-comment-o"
        />
      </button>
      <button
        className="FuseActions-action is-active"
      >
        <i
          aria-hidden="true"
          className="fa fa-heart-o"
        />
      </button>
    </div>
  </article>
</div>
`;

exports[`Fuse renders with an email 1`] = `
<div
  className="Fuse"
>
  <img
    alt="John Smith's Avatar"
    className="Fuse-avatar"
    src="http://avatar.com/me.png"
  />
  <article
    className="Fuse-body"
  >
    <header
      className="Fuse-header"
    >
      <div
        className="Fuse-fullName"
      >
        John Smith
      </div>
      <span>
         
      </span>
      <div
        className="Fuse-userName"
      >
        @
        johnsmith
      </div>
      <span>
         — 
      </span>
      <div
        className="Fuse-date"
        title="October 9 2017 2:00 p.m."
      >
        0 seconds
      </div>
    </header>
    <div
      className="Fuse-message"
    >
      Hello World!
    </div>
    <div
      className="FuseActions"
    >
      <button
        className="FuseActions-action"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="fa fa-comment-o"
        />
      </button>
      <button
        className="FuseActions-action is-active"
      >
        <i
          aria-hidden="true"
          className="fa fa-heart-o"
        />
      </button>
    </div>
  </article>
</div>
`;
